src/main/kotlin/exercise4/task2/TransactionalBankAccount.kt

```diff
@@ -0,0 +1,225 @@
+ *      Amount: [TRANSACTION_AMOUNT]
+ *      Old Balance: [OLD_BALANCE]
+ *      New Balance: [NEW_BALANCE]
+ *      Status: [TRANSACTION_STATUS]
+ *
+ *      Transaction Date: [PRETTY_FORMATED_TRANSACTION_DATE]
+ *      Transaction Type: [TRANSACTION_TYPE]
+ *      Amount: [TRANSACTION_AMOUNT]
+ *      Old Balance: [OLD_BALANCE]
+ *      New Balance: [NEW_BALANCE]
+ *      Status: [TRANSACTION_STATUS]
+ *      ```
+ *     If there is no transaction information should be printed as following:
+ *     ```
+ *      Account Holder: [ACCOUNT_HOLDER_NAME]
+ *      Account Number: [ACCOUNT_NUMBER]
+ *      Balance: [ACCOUNT_BALANCE]
+ *
+ *      Transactions:
+ *
+ *      No transactions recorded.
+ *      ```
+ */
+
+private val currentTime: LocalDateTime get() = LocalDateTime.now()
+
+private fun LocalDateTime.prettyPrint(): String {
+    val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS")
+    return this.format(formatter)
+}
+
+fun main() {
+    // Creating a Transactional Bank Account
+    val account = TransactionalBankAccount("123456789", "John Doe")
+
+    // Making some transactions (successful and unsuccessful)
+    account.deposit(1000.0)
+    account.withdraw(500.0)
+    account.deposit(-100.0)
+    account.withdraw(2000.0)
+    account.deposit(500.0)
+    account.withdraw(100.0)
+
+    // Displaying updated account information
+    account.displayAccountInfo()
+
+    println("All Transactions Dates:")
+    account.getAllTransactions().forEach { println(it.getTransactionDate().prettyPrint()) }
+
+    println("\nTransactions with Amount > 600:")
+    account.getAllTransactionsBy { it.getAmount() > 600 }.forEach{ println(it.getAmount()) }
+
+    println("\nTransactions in the Last Hour:")
+    account.getTransactionsBetween(LocalDateTime.now().minusHours(1), LocalDateTime.now()).forEach { println(it.getAmount()) }
+
+    println("\nAll Successful Transactions:")
+    account.getAllSuccessfulTransactions().forEach { println(it.getAmount()) }
+
+    println("\nAll Failed Transactions:")
+    account.getAllFailedTransactions().forEach { println(it.getAmount()) }
+
+    println("\nAll Successful Deposits:")
+    account.getAllSuccessfulDeposits().forEach { println(it.getAmount()) }
+
+    println("\nAll Successful Withdrawals:")
+    account.getAllSuccessfulWithdrawals().forEach { println(it.getAmount()) }
+
+    println("\nAll Failed Deposits:")
+    account.getAllFailedDeposits().forEach { println(it.getAmount()) }
+
+    println("\nAll Failed Withdrawals:")
+    account.getAllFailedWithdrawals().forEach { println(it.getAmount()) }
+}
+
+class TransactionalBankAccount( accountNumber: String, accountHolderName: String, balance: Double = 0.0):
+    BankAccount(accountNumber, accountHolderName, balance) {
+
+    private var transactions = mutableListOf<Transaction>()
+
+    fun getAllTransactions(): List<Transaction> {
+        transactions.sortByDescending { it.getTransactionDate() }
+        return transactions
```

Comment:
We should always return the copy of `transactions` instead of returning the same reference. 
The problem with this approach is that any client that is using this API can cast the result of the `getAllTransactions` function invocation to a mutable list and modify the internal state of the `TransactionalBankAccount` class, which breaks the encapsulation of the class.

A better approach is to return a new copy by following: 
```kotlin
return transactions.sortedByDescending { it.getTransactionDate() }
```

Here is one more example that explains on how encapsulation is broken and internal state is leaked: 
```kotlin
fun main() {
    val a = A()
    val transactions = a.transactions() as MutableList<Int>
    println(a.transactions()) // prints [1]
    transactions.add(2)
    println(a.transactions()) // prints [1, 2] even though we modified list externally.
}

class A {
    private val transactions = mutableListOf<Int>(1)
    
    fun transactions(): List<Int> = transactions
}
```